### DISCUSSION TITLE
Joint reaction forces

### DISCUSSION THREAD

**Discussion by ma569** [NVIDIA-Omniverse/PhysX#101] [Q&A] [answered]
Hi! I was wondering if it is possible to extract joint reaction from PhysX? From reading the documentation (https://nvidia-omniverse.github.io/PhysX/physx/5.1.2/docs/Articulations.html#compute-joint-forces), it seems like we can get them for articulations. However, I couldn't find if it is possible with joints.

Thank you in advance for reading this and hope to hear from you soon.

**Comment by preist-nvidia** [comment, marked as answer]
You mean when simulating, or for inverse dynamics as your documentation link suggests?

When simulating: We are currently working on a joint force (transmitted force from parent to child link) reporting API that will be released in 5.2.0.

**Reply by ma569** [reply]
Sorry, yes I meant when simulating e.g. here is a simple example system (www.shorturl.at/fqrCV) along with its free body diagram. I would like to extract reaction forces under the action of gravity as the bar rotates around point A

**Reply by preist-nvidia** [reply]
Yes that'll be straightforward with the upcoming new joint force feature in 5.2.0, so I would recommend awaiting the next release.

**Reply by ma569** [reply]
brilliant thanks :)

**Reply by ma569** [reply]
any rough estimates on when 5.2 might be available?

**Reply by preist-nvidia** [reply]
Roughly mid-May.

**Reply by AccessViolationEnjoyer** [reply]
@preist-nvidia Did this make it into 5.2? I couldn't find anything about it in the changelog or documentation.

**Reply by preist-nvidia** [reply]
Yes it did, please have a look here: https://nvidia-omniverse.github.io/PhysX/physx/5.3.0/_api_build/class_px_articulation_cache.html?highlight=link%20incoming%20joint#_CPPv4N19PxArticulationCache22linkIncomingJointForceE

We have an issue with their docs dropping from 5.3.0, but they are in 5.2.1: https://nvidia-omniverse.github.io/PhysX/physx/5.2.1/docs/Articulations.html#articulation-link-sensors

I'm working on fixing the docs asap.

**Reply by preist-nvidia** [reply]
And to be specific about your 5.2 question, yes they were already in then. We improved the TGS behavior of the reported sensor forces when using drives in 5.3.0, though, see the drive comment in the changelog: https://github.com/NVIDIA-Omniverse/PhysX/blob/main/physx/CHANGELOG.md#fixed-2

**Reply by AccessViolationEnjoyer** [reply]
Sorry, perhaps I misunderstood. This appears to be for articulations and I thought the question was asking about regular (maximal coordinate) joints. Is it possible to obtain the force/torque exerted by a drive on a D6Joint acting between two PxRigidDynamic bodies, for example?

**Reply by preist-nvidia** [reply]
No the question was about articulations. But yes, it's possible to get the constraint forces for a D6. You get the constraint pointer using this: https://nvidia-omniverse.github.io/PhysX/physx/5.3.0/_api_build/class_px_d6_joint.html#_CPPv4NK9PxD6Joint13getConstraintEv

And then call the constraint's https://nvidia-omniverse.github.io/PhysX/physx/5.3.0/_api_build/class_px_constraint.html#_CPPv4NK12PxConstraint8getForceER6PxVec3R6PxVec3

Note that we found an issue with the reported force using the TGS solver and nonzero velocity iterations. We are working on a fix for this issue, but I don't have an ETA yet.

**Reply by preist-nvidia** [reply]
@kermado - the docs are fixed, FYI, and the incoming link joint forces are there now: https://nvidia-omniverse.github.io/PhysX/physx/5.3.0/docs/Articulations.html#link-incoming-joint-force

**Reply by AccessViolationEnjoyer** [reply]
Thanks. I did try PxConstraint::getForce(), but it was returning unexpected values and so I assumed that these forces were not for the drive. I was using the TGS solver with velocity iterations, so that explains it. Thanks again!
